src/sage/schemes/elliptic_curves/cm.py

```
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
+        {1: [(-3, 1), (-3, 2), (-3, 3), (-4, 1), (-4, 2), (-7, 1), (-7, 2), (-8, 1), (-11, 1), (-19, 1), (-43, 1)], 2: [(-3, 4), (-4, 3), (-8, 2), (-15, 1), (-20, 1), (-24, 1), (-35, 1), (-40, 1)], 3: [(-11, 2), (-23, 1), (-31, 1)], 4: [(-39, 1)], 5: [(-47, 1)]}
+
     """
+    hmax = Integer(hmax)
+    global hDf_dict
+
+    # Easy case where we have already computed and cached the relevant values
+    if hDf_dict and hmax <= max(hDf_dict.keys()):
+        T = {h:Dflist for h,Dflist in hDf_dict.items() if h<=hmax}
+        if B:
+            for h in T:
+                T[h] = [Df for Df in T[h] if Df[0].abs()*Df[1]**2<=B]
+        return T
+
     # imports that are needed only for this function
+    from sage.arith.srange import xsrange
     from sage.structure.proof.proof import get_flag
-
-    # deal with input defaults and type checking
     proof = get_flag(proof, 'number_field')
-    hmax = Integer(hmax)
-
-    # T stores the output
-    T = {}
-
-    # Easy case -- instead of giving error, give meaningful output
-    if hmax < 1:
-        return T
 
     if B is None:
-        # Determine how far we have to go by applying Watkins's theorem.
-        v = [largest_fundamental_disc_with_class_number(h) for h in range(1, hmax+1)]
+        # Determine how far we have to go by applying Watkins + Klaise's results.
+        v = [largest_disc_with_class_number(h) for h in range(1, hmax+1)]
```

if hmax is larger than 100, it might be better to raise an error telling the user to specify B, rather than a `NotImplementedError`.

---

Now a ValueError will be raised if `hmax>100` and ``B`` is ``None``.